### DISCUSSION TITLE
HDR Gain Map Support

### DISCUSSION THREAD

**Discussion by blue-j** [ImageMagick/ImageMagick#6502] [Development]
### ImageMagick version

7.1.1-13

### Operating system, version and so on

macOS 13.4.1 (c)

### Description

First, trust me, I do not stare into the spinning hypnotic wheel of Adobe nor glorify (or vilify) them.  I understand that "because Adobe does it" is not a good enough reason for supporting a feature.

They have recently come out with an HDR solution that I think is pretty damn cool though.  It may help people take advantage of HDR displays more.  And for once they provide the details needed to really support it.  I ask the leadership here to consider supporting the specification:

[https://helpx.adobe.com/camera-raw/using/gain-map.html](https://helpx.adobe.com/camera-raw/using/gain-map.html)

They provide sample images, a through spec doc, in addition to a small app you can use to visualize results.  Pretty neat stuff!

- J

**Comment by fmw42** [comment]
Gain maps and multi-image (mpo) format are being looked into already.

**Comment by snibgo** [comment]
I agree the Gain Map proposal is good.

I have examined the implementation for JPEG, using MPF (aka MPO). Reading those files is reasonably simple. Writing them is a pain, and I haven't written an implementation. My life is currently too chaotic to give it any time.

I haven't looked at other file formats, eg AVIF and JXL.

The Adobe sample application is good, but it only reads files, and can't create them. It would be useful if they published the source under a FOSS licence that meant their reading code could be copy/pasted into IM.

It occurs to me that there must be software somewhere that can read and write files with gain maps. Is any of that software published? 

EDIT : See also https://github.com/ImageMagick/ImageMagick/discussions/5441

**Comment by blue-j** [comment]
I wrote Eric Chan at Adobe and asked.  Why not?  - J

**Comment by blue-j** [comment]
I heard back from Eric, what a great person.  I'm grateful to him for taking the time.  They can't release the code of the demo app because it is basically Adobe Camera Raw.  They are trying to think of more ideas for supporting the initiative.

He did mention something exciting - "Google has recently implemented the Gain Map in Chromium, and that is also open source. (This means that the sample gain map files on the Adobe page can be viewed in current versions of Chrome Canary.)"

I have not had any luck finding the code in the Chromium repo though, YMMV!

- J

**Comment by snibgo** [comment]
> I have not had any luck finding the code in the Chromium repo though, YMMV!

It seems to be in: https://chromium.googlesource.com/skia/+/76fb1b06bb13db62ea441308a7fb6fa48cd2c9d1/src/codec/SkJpegGainmap.cpp

But, of course, that code calls loads of other code. I'm not sure if that would be useful to copy into IM.

**Comment by gregbenz** [comment]
There is a feature request for gain map support in ImageMagick with a lot of related discussion at https://github.com/ImageMagick/ImageMagick/issues/6377

**Comment by blue-j** [comment]
i'm going to close this discussion and point everyone toward [https://github.com/ImageMagick/ImageMagick/issues/6377](https://github.com/ImageMagick/ImageMagick/issues/6377)
